### PR TITLE
ManagementGroup: Run WhatIf test on e2e after deployment

### PR DESCRIPTION
**Pull request by olljanat**
# Overview/Summary
 Run WhatIf deployment for management group on e2e tests. Also added new optional parameter `parTopLevelManagementGroupParentId` which was needed for it.

## This PR fixes/adds/changes/removes
1. First step to solve #249

### Breaking Changes
N/A

## Testing Evidence
Without `parTopLevelManagementGroupParentId`  parameter new WhatIf test step failed to error and it that it passed without issues.
```
Note: The result may contain false positive predictions (noise).
You can help us improve the accuracy of the result by opening an issue here: https://aka.ms/WhatIfIssues

Resource and property changes are indicated with these symbols:
  - Delete
  ~ Modify

The deployment will update the following scope:

Scope: /

  ~ Microsoft.Management/ManagementGroups/PR-test [2021-04-01]
    - properties.details:

        parent.id: "/providers/Microsoft.Management/managementGroups/13a69a3b-cf5f-4204-b274-3e9ce5240a60"


Resource changes: 1 to modify.
```

## As part of this Pull Request I have

- [x] Read the [Contribution Guide](https://github.com/Azure/ALZ-Bicep/wiki/Contributing) and ensured this PR is compliant with the guide
- [x] Checked for duplicate [Pull Requests](https://github.com/Azure/ALZ-Bicep/pulls)
- [x] Associated it with relevant [GitHub Issues](https://github.com/Azure/ALZ-Bicep/issues)
- [ ] *(ALZ Bicep Core Team Only)* Associated it with relevant [ADO Items](https://aka.ms/alz/bicep/backlog)
- [x] Ensured my code/branch is up-to-date with the latest changes in the `main` [branch](https://github.com/Azure/ALZ-Bicep/tree/main)
- [x] Performed testing and provided evidence.
- [x] Updated tests *(if required)* [Unit](https://github.com/Azure/ALZ-Bicep/blob/main/.github/workflows/bicep-build-to-validate.yml) - [Linting](https://github.com/Azure/ALZ-Bicep/tree/main/.github/workflows) - [E2E (End-To-End)](https://github.com/Azure/ALZ-Bicep/blob/main/tests/pipelines/bicep-build-to-validate.yml)
- [ ] Updated relevant and associated documentation (e.g. Contribution Guide, Module READMEs, Wiki Docs etc.)
- [ ] If relevant, created or updated Code Tours [here](https://github.com/Azure/ALZ-Bicep/blob/main/.vscode/tours)

### CODE REVIEW

**Comment by jtracey93** [comment]
Thanks @olljanat for this work. As mentioned yesterday (in another discussion IIRC) we are not merging any other changes until #227 is complete and merged as this is a big PR and we need to close it out first before we keep making more merge conflicts we need to resolve.

Can I propose we put this on hold until #227 is merged (hopefully next week) and then you can update this PR with the latest from the `upstream` (this repo) and then we can review and take it from there?

Let me know if all okay with that 👍

**Comment by olljanat** [comment]
@jtracey93 that's fine. I added my self to #227 notification list and will rebase my PRs after it is merged.

**Comment by olljanat** [comment]
@jtracey93 rebased with main

**Comment by olljanat** [comment]
@jtracey93 updated "all" parameters file and rebased with main.

File: tests/pipelines/bicep-build-to-validate.yml
Comment (jtracey93) [inline tests/pipelines/bicep-build-to-validate.yml:257]: @olljanat the `--parameters` file path is incorrect here. Can you update?

File: tests/pipelines/bicep-build-to-validate.yml
Comment (jtracey93) [inline tests/pipelines/bicep-build-to-validate.yml:256]: @olljanat why the `test` part in here?

File: infra-as-code/bicep/modules/managementGroups/managementGroups.bicep
Comment (jtracey93) [inline infra-as-code/bicep/modules/managementGroups/managementGroups.bicep:4]: @olljanat should we just set the default value to `tenant().tenantId`?

This way all customers by default will pass the what-if stage?

**Review by jtracey93 — CHANGES_REQUESTED** [review]
@olljanat a few more bits.

Also requesting @jfaurskov to review this PR too

**Review by jfaurskov — COMMENTED** [review]
Looking good to me. We should expand on this going forward as we split validation and deployment tests up into individual chunks.

**Comment by jtracey93** [comment]
> Looking good to me. We should expand on this going forward as we split validation and deployment tests up into individual chunks.

Agreed and think these should be included on the deployment phase as needs to run again post deployment as a check that the what-if doesn't kick off saying there's a change, when there clearly wouldn't be 👍

File: tests/pipelines/bicep-build-to-validate.yml
Comment (olljanat) [reply tests/pipelines/bicep-build-to-validate.yml:256]: Perhaps because I'm not that familiar with jq syntax and have copied some example from internet 😄

Comment (jtracey93) [reply tests/pipelines/bicep-build-to-validate.yml:256]: Here is what i would use instead `tenantRootGroupID=$(az account management-group list | jq '.[] | select(.displayName=="Tenant Root Group") | .id' | tr -d '"')`

File: infra-as-code/bicep/modules/managementGroups/managementGroups.bicep
Comment (jtracey93) [inline infra-as-code/bicep/modules/managementGroups/managementGroups.bicep:4]:
```suggestion
param parTopLevelManagementGroupParentId string = '/providers/Microsoft.Management/managementGroups/${tenant().tenantId}'
```

Comment (jtracey93) [reply infra-as-code/bicep/modules/managementGroups/managementGroups.bicep:4]: Have provided suggestion that I have tested

File: tests/pipelines/bicep-build-to-validate.yml
Comment (jtracey93) [reply tests/pipelines/bicep-build-to-validate.yml:256]:
```suggestion
        tenantRootGroupID=$(az account management-group list | jq '.[] | select(.displayName=="Tenant Root Group") | .id' | tr -d '"')
```

**Comment by olljanat** [comment]
@jtracey93 looks that your idea to use `tenant().tenantId` handles the WhatIf case nicely so there is no even need to have separate parameter at all \o/

**Comment by jtracey93** [comment]
> @jtracey93 looks that your idea to use `tenant().tenantId` handles the WhatIf case nicely so there is no even need to have separate parameter at all \o/

Hey @olljanat I actually think having this as an input parameter is a nice idea as it enables some further use cases for nested ALZ deployments in some more complex scenarios 👍

**Comment by olljanat** [comment]
@jtracey93 OK. Added it back but updated description because it is not really needed by WhatIf anymore (and not used in e2e because of that).

File: infra-as-code/bicep/modules/managementGroups/managementGroups.bicep
Comment (jtracey93) [inline infra-as-code/bicep/modules/managementGroups/managementGroups.bicep:77]: Can we add the logic back here that if an empty string ` ` is provided that we pass in `/providers/Microsoft.Management/managementGroups/${tenant().tenantId}`?

If any other string is provided im happy for the error, like our test just picked up 😄 
```
ERROR: {"status":"Failed","error":{"code":"DeploymentFailed","message":"At least one resource deployment operation failed. Please list deployment operations for details. Please see https://aka.ms/DeployOperations for usage details.","details":[{"code":"BadRequest","message":"{\r\n  \"error\": {\r\n    \"code\": \"BadRequest\",\r\n    \"message\": \"Invalid management group creation parameters: expected 'parentId' format: '/providers/Microsoft.Management/managementGroups/ID'\",\r\n    \"details\": null\r\n  }\r\n}"}]}}
```

If you can make the change, then we can get this one merged 👍

File: infra-as-code/bicep/modules/managementGroups/README.md
Comment (jtracey93) [inline infra-as-code/bicep/modules/managementGroups/README.md:23]: Can we add this to the example column for this please to help show what is expected: `/providers/Microsoft.Management/managementGroups/xxxxxxxx-xxxx-xxxx-xxxx-xxxxxxxxxxxx`

Thanks 👍

**Review by jtracey93 — CHANGES_REQUESTED** [review]
Couple of changes to make it more robust for all users @olljanat (thanks for your patience)

**Comment by jtracey93** [comment]
Closing this as have wrapped into #276

**Pull request closed without merging**
